keras/layers/crf.py

```diff
@@ -0,0 +1,220 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+from .. import backend as K
+from .. import initializations, regularizers, constraints
+from ..engine import Layer, InputSpec
+
+
+def sparse_chain_crf_loss(y, x, U, b):
+    '''Given the true sparsely encoded tag sequences y, observations x,
+    transition params U and label bias b, it computes the loss function
+    of a Linear Chain Conditional Random Field.
+    '''
+    energy = path_energy(y, x, U, b)
+    energy -= free_energy(x, U, b)
+    return -energy
+
+
+def chain_crf_loss(y, x, U, b):
+    '''Given the true tag sequences y as one-hot encoded vectors,
+    observations x, transition params U and label bias b, it computes the loss
+    function of a Linear Chain Conditional Random Field.
+    '''
+    y_sparse = K.argmax(y, -1)
+    y_sparse = K.cast(y_sparse, 'int32')
+    return sparse_chain_crf_loss(y_sparse, x, U, b)
+
+
+def viterbi_decode(x, U, b):
+
+    def _forward_step(x_t, states):
+        gamma_tm1, alpha_tm1, U_shared = states
+        B = K.expand_dims(alpha_tm1, 2) + K.expand_dims(x_t, 1) + K.expand_dims(U_shared, 0)
```

Comment:
With not mask, how do you handle a batch with different sequence lengths? Said a batch of 2, with first input have lenght 2 and the 2nd has a lenght 4. Then you would do a right padding on the 1st sequence to make it have length 4. In you implementation without mask, for the 1st input, the **true** energy should be 
`b_start + x1' y1 + x2' y2 + y1' U y2 + b_end`
but in your implementation, there isn't a `b_end`, but with an additional
`x3' y3' + x4' y4 + y2' U y3 + y3' U y4`,
where `y3 = y4 = 0`. 
The consequence is, the above two formulations are not equivalent, at least when you take derivative with respect to `U_00`, the derivative isn't the same. Right?

Also when you compute the normalization constance (free energy in your code), you have to integrate over `y3, y4` (which are paddings). I guess that's what you mean by "padding elements act as a virtual end label". However, if when you think about taking derivative with respect to `U` or `b_end`, your approach is not equivalent to a real CRF. 

One very obvious observation is, `y3, y4`, the padding, affects the derivative with respect to `U`,  and therefore, the paddings plays a role on the final outcome. The more paddings you have, the more impact the paddings affects the outcome. This is unexpected from my point of view.

Lastly, another simple observation, a model with and without the end energy (`b_end`), the numbers of trainable parameters are not the same. So the two models are not the same.